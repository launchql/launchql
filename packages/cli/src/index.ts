#!/usr/bin/env node
import { CLI, CLIOptions } from 'inquirerer';

import { commands } from './commands';

export const options: Partial<CLIOptions> = {
  minimistOpts: {
    alias: {
      v: 'version',
      h: 'help'
    }
  }
};

const app = new CLI(commands, options);

app.run().then(()=> {
  // all done!
}).catch(error => {
  // Should not reach here with the new CLI error handling pattern
  // But keep as fallback for unexpected errors
  console.error('Unexpected error:', error);
  process.exit(1);
});