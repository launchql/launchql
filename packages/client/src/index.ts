import { Pool, PoolClient } from 'pg';

import env from './env';

export class Database {
  private static instance: Database;
  private pool: Pool;

  constructor() {
    if (Database.instance) {
      return Database.instance;
    }

    const pgPoolConfig = {
      connectionString: env.DATABASE_URL,
    };
    this.pool = new Pool(pgPoolConfig);

    // Ensure the pool is closed on process termination
    process.on('SIGTERM', async () => {
      await this.shutdown();
    });

    Database.instance = this;
    return this;
  }

  /**
   * Executes a callback function within a database transaction.
   * @param fn - A callback function that receives a PoolClient to perform database operations.
   */
  async withTransaction(fn: (client: PoolClient) => Promise<void>): Promise<void> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      try {
        await fn(client);
        await client.query('COMMIT');
      } catch (e) {
        console.error('Error during transaction:', e);
        await client.query('ROLLBACK');
        throw e; // Re-throw the error to propagate it
      }
    } finally {
      client.release();
    }
  }

  /**
   * Shuts down the connection pool.
   */
  async shutdown(): Promise<void> {
    await this.pool.end();
  }
}