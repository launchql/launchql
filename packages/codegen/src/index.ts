import { generateCode } from './codegen';
import getIntrospectionRows, { GetIntrospectionRowsOptions } from './introspect';
import { DatabaseObject } from './types';

(async () => {
  const options: GetIntrospectionRowsOptions = {
    connectionString: 'postgresql://postgres:password@localhost:5432/mydb',
    introspectionOptions: {
      pgLegacyFunctionsOnly: false,
      pgIgnoreRBAC: true,
    },
    namespacesToIntrospect: ['collections_public'],
    includeExtensions: false,
  };

  try {
    // Fetch introspection rows
    const rows: DatabaseObject[] = await getIntrospectionRows(options);
    console.log('Introspection Rows Fetched:', rows);

    // Generate TypeScript code
    const codegenOptions = {
      includeTimestamps: true,
      includeUUID: true,
    };
    const generatedCode = generateCode(rows, codegenOptions);

    console.log('Generated TypeScript Code:');
    console.log(generatedCode);
  } catch (error) {
    console.error('Failed to fetch introspection rows or generate code:', error);
  }
})();
