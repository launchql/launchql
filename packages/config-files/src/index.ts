export * from './types';
export * from './plan';
export * from './config';
export * from './sql';
export * from './extension';
export * from './sql-scripts';