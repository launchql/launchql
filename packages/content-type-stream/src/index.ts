export * from './content-stream';
export * from './content-type-stream';