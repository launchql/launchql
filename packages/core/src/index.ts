export * from './core/class/launchql';
export * from './export/export-meta';
export * from './export/export-migrations';
export * from './extensions/extensions';
export * from './modules/modules';
export * from './packaging/package';
export * from './packaging/transform';
export * from './resolution/deps';
export * from './resolution/resolve';
export * from './workspace/paths';
export * from './workspace/utils';

// Export project-files functionality (now integrated into core)
export * from './files';
export { cleanSql } from './migrate/clean';
export { LaunchQLMigrate } from './migrate/client';
export { 
  DeployOptions, 
  DeployResult, 
  MigrateChange, 
  MigratePlanFile, 
  RevertOptions, 
  RevertResult, 
  StatusResult,
  VerifyOptions, 
  VerifyResult} from './migrate/types';
export { hashFile, hashString } from './migrate/utils/hash';
export { executeQuery,TransactionContext, TransactionOptions, withTransaction } from './migrate/utils/transaction';
