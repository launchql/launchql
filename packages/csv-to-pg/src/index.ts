// @ts-nocheck
export * from './parse';
export * from './parser';
