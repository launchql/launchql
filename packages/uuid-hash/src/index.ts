import { createHash as createCryptoHash, Hash as NodeHash } from 'crypto';

export const DNS = '6ba7b810-9dad-11d1-80b4-00c04fd430c8';
export const URL = '6ba7b811-9dad-11d1-80b4-00c04fd430c8';

const convertToBuffer = (bytes: string | number[] | Buffer): Buffer => {
  if (Array.isArray(bytes)) {
    return Buffer.from(bytes);
  } else if (typeof bytes === 'string') {
    return Buffer.from(bytes, 'utf8');
  }
  return bytes;
};

const uuidToBytes = (uuid: string): number[] => {
  const bytes: number[] = [];
  uuid.replace(/[a-fA-F0-9]{2}/g, (hex) => {
    bytes.push(parseInt(hex, 16));
    return '';
  });
  return bytes;
};

const hex: string[] = [];
for (let i = 0; i < 256; i++) {
  hex[i] = (i < 16 ? '0' : '') + i.toString(16);
}

const hexToBytes = (hexStr: string): number[] => {
  const bytes: number[] = [];
  for (let c = 0; c < hexStr.length; c += 2) {
    bytes.push(parseInt(hexStr.substr(c, 2), 16));
  }
  return bytes;
};

export class UuidHash {
  private version: number;
  private namespace: string | number[];
  private shasum: NodeHash;

  constructor(namespace: string | number[] = URL, version: number = 0x50) {
    this.version = version;
    this.namespace = namespace;

    this.shasum = createCryptoHash('sha1');
    this.shasum.update(convertToBuffer(this.namespace));

    if (typeof this.namespace === 'string') {
      this.namespace = uuidToBytes(this.namespace);
    }

    if (!Array.isArray(this.namespace) || this.namespace.length !== 16) {
      throw new Error('namespace must be uuid string or an Array of 16 byte values');
    }
  }

  update(chunk: string | number[] | Buffer): this {
    this.shasum.update(convertToBuffer(chunk));
    return this;
  }

  digest(): string {
    const r = hexToBytes(this.shasum.digest('hex'));
    r[6] = (r[6] & 0x0f) | this.version;
    r[8] = (r[8] & 0x3f) | 0x80;

    return (
      hex[r[0]] +
      hex[r[1]] +
      hex[r[2]] +
      hex[r[3]] +
      '-' +
      hex[r[4]] +
      hex[r[5]] +
      '-' +
      hex[r[6]] +
      hex[r[7]] +
      '-' +
      hex[r[8]] +
      hex[r[9]] +
      '-' +
      hex[r[10]] +
      hex[r[11]] +
      hex[r[12]] +
      hex[r[13]] +
      hex[r[14]] +
      hex[r[15]]
    );
  }
}

export const createHash = (...args: ConstructorParameters<typeof UuidHash>): UuidHash => {
  return new UuidHash(...args);
};

export default UuidHash;