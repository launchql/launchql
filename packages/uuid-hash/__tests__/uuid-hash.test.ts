import { createHash } from '../src';

describe('UUID v5', () => {
  it('uuid v5 hash', async () => {
    const res: Record<string, string> = {};
    const strings = ['Hello World', 'Another String', 'uuid'];
    for (let i = 0; i < strings.length; i++) {
      const str = strings[i];
      res[str] = createHash()
        .update(str)
        .digest();
    }
    expect(res).toMatchSnapshot();
  });

  it('uuid v5 hash w custom namespace', async () => {
    const res: Record<string, string> = {};
    const strings = ['Hello World', 'Another String', 'uuid'];
    for (let i = 0; i < strings.length; i++) {
      const str = strings[i];
      res[str] = createHash('e8613ca4-b17d-5979-82ea-86c1373c4ffc')
        .update(str)
        .digest();
    }
    expect(res).toMatchSnapshot();
  });
});
