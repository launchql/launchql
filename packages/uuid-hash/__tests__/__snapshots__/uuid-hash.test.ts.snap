// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UUID v5 uuid v5 hash 1`] = `
{
  "Another String": "6683e2b8-dbf7-5927-bdf0-22520ace81b1",
  "Hello World": "7bf3746c-6273-5e2d-8174-be6c0a78c192",
  "uuid": "e8613ca4-b17d-5979-82ea-86c1373c4ffc",
}
`;

exports[`UUID v5 uuid v5 hash w custom namespace 1`] = `
{
  "Another String": "1afb8e5d-2c55-5c54-b596-741f19965158",
  "Hello World": "a17fe188-1546-54ac-8e87-6a58931ea342",
  "uuid": "2107d0c4-2b6c-5583-90f4-ee0b303c3a58",
}
`;
